src/Grpc.AspNetCore.Server/Internal/PipeExtensions.cs

```diff
@@ -0,0 +1,262 @@
+            else if (flag == 1)
+            {
+                return true;
+            }
+            else
+            {
+                throw new InvalidOperationException("Unexpected compressed flag value in message header.");
+            }
+        }
+
+        /// <summary>
+        /// Read a single message from the pipe reader. Ensure the reader completes without additonal data.
+        /// </summary>
+        /// <param name="input">The request pipe reader.</param>
+        /// <returns>Complete message data.</returns>
+        public static async ValueTask<byte[]> ReadMessageAsync(this PipeReader input)
+        {
+            byte[] completeMessageData = null;
+
+            while (true)
+            {
+                var result = await input.ReadAsync();
+                var buffer = result.Buffer;
+
+                try
+                {
+                    if (result.IsCanceled)
+                    {
+                        throw new InvalidOperationException("Incoming message cancelled.");
+                    }
+
+                    if (!buffer.IsEmpty)
+                    {
+                        if (completeMessageData != null)
+                        {
+                            throw new InvalidOperationException("Additional data after the message received.");
+                        }
+
+                        if (TryReadMessage(ref buffer, out var data))
+                        {
+                            // Store the message data
+                            // Need to verify the request completes with no additional data
+                            completeMessageData = data;
+                        }
+                    }
+
+                    if (result.IsCompleted)
+                    {
+                        if (completeMessageData != null)
+                        {
+                            // Finished and the complete message has arrived
+                            return completeMessageData;
+                        }
+
+                        throw new InvalidOperationException("Incomplete message.");
+                    }
+                }
+                finally
+                {
+                    // The buffer was sliced up to where it was consumed, so we can just advance to the start.
+                    // We mark examined as buffer.End so that if we didn't receive a full frame, we'll wait for more data
+                    // before yielding the read again.
+                    input.AdvanceTo(buffer.Start, buffer.End);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Read a message in a stream from the pipe reader. Additional message data is left in the reader.
+        /// </summary>
+        /// <param name="input">The request pipe reader.</param>
+        /// <returns>Complete message data or null if the stream is complete.</returns>
+        public static async ValueTask<byte[]> ReadMessageStreamAsync(this PipeReader input)
```

Comment:
I'd like to see some numbers with IAsyncEnumerable vs the current method before we change it. How would it make it faster? As far as I know it would add overhead of allocating an enumerable. We have an IAsyncStreamReader for that purpose already.

Reply:
- We would write a single loop which means a single async state machine for the entire enumeration, not one per call.
- We would implement the IAsyncStreamReader by getting the enumerator and by manually calling MoveNext() and Current on it.

Reply:
Regardless of what ReadMessageStreamAsync() returns,  shouldn't we deduplicate the logic for reading (that's what my original comment was about)?
My suggestion is to make the IAsyncEnumerable switch in a separate PR (where we can collect performance data if needed).